### ISSUE TITLE
HDR support in Hyprland

### ISSUE THREAD

**Issue by Swomp2** [hyprwm/Hyprland#4933]
### Description

So, plasma 6 was recently released and has initial HDR support, which, as far as I know, can be ported to other graphics environments.  GNOME, according to rumors, will also soon receive HDR support.  Do you have any future plans to bring HDR to hyprland?

**Comment by vaxerski** [comment]
No; I don't have an HDR display. Feel free to MR, though, nothing against it.

**Comment by Swomp2** [comment]
Okay, thank you

**Comment by nergdron** [comment]
what's MR stand for? also if someone contributed an HDR display to you, would this at all change the priority for you?

**Comment by vaxerski** [comment]
merge request.

And please dont contribute an hdr display to me.

**Comment by nergdron** [comment]
haha ok well. I'm not a C++ dev, so I'm not really able to create a patch to add this support, but it is super important to me. if there's something that would be useful to you that'd encourage you to work on it, just say the word.

**Comment by vaxerski** [comment]
I looked through the specs of my laptop and it _should_ support HDR?

![image](https://github.com/hyprwm/Hyprland/assets/43317083/8b719236-0624-4f6c-afe1-bf7cfa5136ce)

Anyways, we still should wait for the w-p, KDE uses a vendored, custom protocol it seems.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/14

Also why was this closed? It hasn't been solved?

I can dig into HDR a bit if my laptop proves to be actually compatible with HDR

**Comment by crimist** [comment]
> I looked through the specs of my laptop and it should support HDR?

Yes, those specs indicate your laptop has an OLED panel which will deliver an excellent HDR experience. It'll work for testing.

@Swomp2 It'd be great if you could reopen this issue to track progress.

**Comment by crimist** [comment]
Regarding the general state of HDR on Linux things are still in their infancy. @Zamundaaa has great [blog posts](https://zamundaaa.github.io/wayland/2023/12/18/update-on-hdr-and-colormanagement-in-plasma.html#fn:4) about it but the TLDR is that most things aren't finalized or merged yet.

I believe any WM can implement HDR in a way that works with a hacky Vulkan layer (like Plasma) but it may have to evolve a bit when full support comes.

I'm far from an expert so any and all corrections are welcome.

**Comment by vaxerski** [comment]
https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4594

**Comment by Swomp2** [comment]
I closed the issue because I thought no one would be interested in it + personally, I’m not a programmer and I won’t be able to add anything ((

**Comment by vaxerski** [comment]
still it's better to have a place to track it

**Comment by abysssdweller** [comment]
HDR on hyprland would be crazy!

**Comment by ysimard94** [comment]
Lack of HDR support is the only reason why I'm not making the full switch from KDE, would love for this to happen

**Comment by crimist** [comment]
The enthusiasm is nice but HDR won't be implemented until the wlroots MR is merged.
If you're looking to show your support give the issue a thumbs up instead so that subscribers can track progress.

**Comment by abihf** [comment]
Btw, because Hyprland ship with its own wlroots, is it possible to patch wlroots with that MR?

@vaxerski I'm not familiar with color transformation, is there anything I can help?

My suggestion is first we enable hdr metadata on compositor level with sdr -> hdr tone mapping, next step will be protocol implementation so client can show their hdr contents.

**Comment by vaxerski** [comment]
we don't patch wlroots at all, we don't maintain a fork either. So, no.

**Comment by Andorreta** [comment]
Is there a way to enable our custom wlroots? Because when I try to enable the Vulkan renderer in my multi head setup, everything breaks 😆 And to use the Wayland HDR we need Vulkan (or at least trying with gamescope)

**Comment by vaxerski** [comment]
> Is there a way to enable our custom wlroots?

No.

> Because when I try to enable the Vulkan renderer in my multi head setup, everything breaks 

Hyprland doesn't support vulkan as a backend

> And to use the Wayland HDR we need Vulkan (or at least trying with gamescope)

This is false.

**Comment by Andorreta** [comment]
Don't know if vilkan is really needed if using wlroots directly. But acording to [this post](https://zamundaaa.github.io/wayland/2023/12/18/update-on-hdr-and-colormanagement-in-plasma.html) from a KDE developer, you need to install a vulkan layer to be able to use HDR to play the games in Kwin6 using HDR :smile: 

If we (or i) can help in something, i could help code or reserach a bit more to see what's needed.
The ENV variable that breaks Hyprland for me is `env = WLR_RENDERER,vulkan` . But without it i can't enable HDR in gamescope. If someone knows a way to enable HDR in gamescope using hyprland, please tell me :smile:

**Comment by cajual** [comment]
After reading this I'm wondering if I am missing something integral to the conversation.

1. Up until today Hyprland was running in typical SDR, as expected. I use HighRes and 10-bit for my monitor, but otherwise it was showing as 3840x2160@120 VRR RGB 10b 4L10 SDR.
2. SDDM was causing me issues and going to "no input" with X11, so I switched to Wayland and copied over my `kwinoutputconfig.json` to my SDDM config.
3. SDDM now shows up on my screen without issue, but when I log-in to Hyprland, my monitor immediately shows the HDR logo and I see 3840x2160@120 VRR RGB 10b 4L10 HDR10.
4. The color profile is pretty intense and RGB aliasing is pretty rough at 1:1 scaling, so now I am wondering why Hyprland is showing as HDR.
5. Disabling Wayland HDR support from Plasma returns Hyprland to SDR.

**Comment by vaxerski** [comment]
kek drm did something funny.

I assume then that SDDM was running with KDE as a backend which supports HDR. Setting an HDR "mode" is the simple part of HDR and KWin probably submitted that to the pipeline

Hyprland doesn't support the HDR "mode" hints and thus can't change it.

That's my guess. Tho I have very little knowledge about HDR internals

**Comment by cajual** [comment]
I fixed it by strictly disabling HDR and Wide Color Gamut in the SDDM kwinoutputconfig.json file. This means there isn't any HDR in the pipeline for Hyprland. Finally have proper colors, and it doesn't affect my Plasma session HDR.

**Comment by ous50** [comment]
> kek drm did something funny.
> 
> I assume then that SDDM was running with KDE as a backend which supports HDR. Setting an HDR "mode" is the simple part of HDR and KWin probably submitted that to the pipeline
> 
> Hyprland doesn't support the HDR "mode" hints and thus can't change it.
> 
> That's my guess. Tho I have very little knowledge about HDR internals

Enabling HDR in Plasmas 6 requires setting the following envrionment variables:

```shell
export KWIN_DRM_ALLOW_INTEL_COLORSPACE=1
```

**Comment by ducviet00** [comment]
Is there any update on this? That would be great if Hyprland supports HDR because I just ordered an HDR monitor (jk lol)

**Comment by vaxerski** [comment]
not atm

**Comment by crimist** [comment]
For subscribers: I believe HDR is no longer blocked on upstream wlroots since Hyprland migrated to aquamarine in #6608.
If you'd like to contribute HDR related features you can probably do so at https://github.com/hyprwm/aquamarine?

**Comment by vaxerski** [comment]
correct, though worth noting it will require support in both aq and hl.

The first (easy) step would be to implement what KDE did - just the hdr profiles (not the actual data, IIRC that's still blocked on w-p mesa and kernel)

**Comment by Swomp2** [comment]
WOW!! 
I didn't know that you wrote new library, it's incredible work and I still hope that Hyprland will support HDR. Thank you all for your work, guys

**Comment by Ciel-MC** [comment]
The only way I play HDR now is by switching to gamescope session, which has... quite a few inconveniences. Please call me when/if ever it's time for testing, thanks. (Or if there is work to be done I suppose, I can try but don't expect my C++ skills :p)

**Comment by proJM-Coding** [comment]
Sadly gamescope is not a solution for everyone as it doesn't work well with NVIDIA gpus. The gamescope session also doesn't work out of the box if you are not using systemd.

**Comment by littleblack111** [comment]
> correct, though worth noting it will require support in both aq and hl.

aren't those both owned by hyprwm? why can't we be like the KDE devs and implement it...

also do we need that protocol? Like other wayland protocols, will take forever to merge 😭

**Comment by vaxerski** [comment]
> why can't we be like the KDE devs and implement it...

because KDE has 50 full-time paid developers

**Comment by littleblack111** [comment]
true...

**Comment by fxzzi** [comment]
this can be closed now ja?

**Comment by vaxerski** [comment]
I believe so. Shoutout @UjinT34 for writing the support.

**Comment by MikeWalrus** [comment]
I think closing this might be a bit misleading. Can we keep this open until at least the color management code related to HDR is written?

**Comment by vaxerski** [comment]
uhh, #8715 is merged

**Comment by MikeWalrus** [comment]
Full screen HDR10 might technically not need color management, but if the HDR content is not full screen, Hyprland needs to do the color management properly.
